draft-ietf-quic-invariants.md

```diff
@@ -114,10 +112,53 @@ endpoints use QUIC packets to establish a QUIC connection, which is shared
 protocol state between those endpoints.
 
 
+# Notational Conventions
+
+Packet diagrams in this document use a format defined in {{QUIC-TRANSPORT}} to
+illustrate the order and size of fields.
+
+Complex fields are named and then followed by a list of fields surrounded by a
+pair of matching braces. Each field in this list is separated by commas.
+
+Individual fields include length information, plus indications about fixed
+value, optionality, or repetitions. Individual fields use the following
+notational conventions, with all lengths in bits:
+
+x (A):
+: Indicates that x is A bits long
+
+x (A..B):
+: Indicates that x can be any length from A to B; A can be omitted to indicate
+  a minimum of zero bits and B can be omitted to indicate no set upper limit;
+  values in this format always end on an octet boundary
+
+x (?) = C:
+: Indicates that x has a fixed value of C
+
+x (E) ...:
+: Indicates that x is repeated zero or more times (and that each instance is
+  length E)
+
+{{fig-ex-format}} shows an example structure:
+
+~~~
+Example Structure {
+  One-bit Field (1),
+  7-bit Field with Fixed Value (7) = 61,
+  Arbitrary-Length Field (..),
+  Variable-Length Field (8..24),
+  Repeated Field (8) ...,
+}
+~~~
+{: #fig-ex-format title="Example Format"}
+
+
 # QUIC Packet Headers
 
-A QUIC packet is the content of the UDP datagrams exchanged by QUIC endpoints.
-This document describes the contents of those datagrams.
+QUIC endpoints exchange UDP datagrams that contain one or more QUIC packets.
+This section describes the invariant characteristics of a QUIC packet.  A
+version of QUIC could permit multiple QUIC packets in a single UDP datagram, but
+the invariant properties only describe the first packet in a datagram.
```

Comment:
I think it's a design change. I think this allows for some freedom in the future on the part of endhost stack developers, so I like it, but I'm not sure I can simply justify it with that rationale.

That said, the way I read it, a new version of QUIC _could_ have a packet_len field, but the following packet within the same dgram is then not guaranteed to follow the invariants. This seems to deviate from the previous simple model about packets. Why can this not apply to all packets, even if it cannot be enforced?

Reply:
I think that this is far more pragmatic than that.  It says that this document can only talk about the first packet in a datagram because it doesn't have any way of identifying other packets.  That's pure mathematics from my perspective, not a design change.  

I don't see this as anything more than editorial on that basis, but I will pass this to the list.

(BTW, you don't strictly need a length, because you could start from the other end of the datagram payload to find a second packet.)

Reply:
So if we go down this path I think we need carefully to consider each transport mechanism. And the fate of multiplexed flows start to become entangled, such as when a sender chooses a smaller size or DSCP marking because that endpoint expects a characteristic treatment; the unit of loss can be different to the unit of retransmission; PMTUD will need tighter language; pacing may be impacted;  - basically it makes the transport more complex, if that's needed let's try and explain why and understand the constraints - e.g. do we need rules to say when not to multiplex multiple QUIC packets into one IP packet?

Reply:
I don't understand what you mean by "if" here.  This is a choice already made for QUICv1: it has multiple packets in a datagram.  That those packets have most of the invariant properties doesn't matter because that isn't detectable.

So if the listed problems are problems - and I have no idea how - then they are already problems.

Reply:
I agree with @martinthomson here. Since the long header length is version-specific, it's not possible to access any subsequent packets in the same UDP payload from just the invariants. Stating in the invariants that some QUIC versions might allow packet coalescing is not a design change, it's just a fact.

Reply:
I'll take back my suggestion that this should be a design issue. I agree that this new text is consistent with what is possible. (To answer myself, a new version of QUIC could just as well define coalescing as using "sub packets" with other headers, so there's no way to require a header format for a second packet in a coalesced datagram.)

Reply:
And the points I raised were not "problems" - I will take them as things that I think need to be mentioned in the transport ID. I'll check that next time I review.